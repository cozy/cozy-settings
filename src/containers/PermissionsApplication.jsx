import React from 'react'

import { APPS_DOCTYPE } from 'doctypes'
import { Q, useQuery, isQueryLoading, hasQueryBeenLoaded } from 'cozy-client'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Typography from 'cozy-ui/transpiled/react/Typography'
import { useI18n } from 'cozy-ui/transpiled/react'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import { withRouter } from 'react-router-dom'

const PermissionsApplication = ({ match }) => {
  const { t } = useI18n()
  const appName = match.params.app

  const queryResult = useQuery(
    Q(APPS_DOCTYPE).getById('io.cozy.apps/' + appName),
    {
      as: 'io.cozy.apps/' + appName
    }
  )
  console.log(queryResult)

  return (
    <Page narrow>
      {isQueryLoading(queryResult) && !hasQueryBeenLoaded(queryResult) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResult.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <div>
          <PageTitle>{appName.toUpperCase()}</PageTitle>
          {Object.entries(queryResult.data[0].attributes.permissions).map(
            ([key, value]) => (
              <React.Fragment key={key}>
                <Typography variant="h4">
                  {value.type} : {value.verbs ? value.verbs.join(' / ') : 'ALL'}
                </Typography>
              </React.Fragment>
            )
          )}
        </div>
      )}
    </Page>
  )
}

export default withRouter(PermissionsApplication)
