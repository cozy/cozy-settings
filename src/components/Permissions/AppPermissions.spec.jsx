import { render } from '@testing-library/react'
import React from 'react'
import PermissionsApplication, { completePermission } from './AppPermissions'
import {
  Q,
  useQuery,
  useClient,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'

jest.mock('cozy-ui/transpiled/react/I18n/withLocales', () => {
  return () => Component => {
    const t = text => text
    const match = { params: { slug: 'Drive' } }
    // eslint-disable-next-line react/display-name
    return () => <Component match={match} t={t} />
  }
})

jest.mock('cozy-ui/transpiled/react/NavigationList', () => {
  return {
    __esModule: true,
    default: ({ children }) => (
      <div data-testid="NavigationList">{children}</div>
    ),
    NavigationListSection: ({ children }) => (
      <div data-testid="NavigationListSection">{children}</div>
    )
  }
})

jest.mock('react-router-dom', () => {
  return {
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({ slug: 'Drive' }),
    Link: ({ narrow, children }) => (
      <div data-testid="page" data-narrow={narrow}>
        {children}
      </div>
    )
  }
})

jest.mock('cozy-ui/transpiled/react/CozyDialogs', () => {
  return {
    Dialog: () => (
      <div
        data-testid="Dialog"
        data-open="open"
        data-onClose="onClose"
        data-title="title"
        data-content="content"
      />
    )
  }
})

jest.mock('cozy-client', () => ({
  ...jest.requireActual('cozy-client'),
  hasQueryBeenLoaded: jest.fn(),
  Q: jest.fn(),
  useQuery: jest.fn(),
  isQueryLoading: jest.fn(),
  useClient: jest.fn()
}))

jest.mock('components/Page', () => {
  return ({ narrow, children }) => (
    <div data-testid="page" data-narrow={narrow}>
      {children}
    </div>
  )
})

jest.mock('components/PageTitle', () => {
  return ({ children }) => <div data-testid="pageTitle">{children}</div>
})
jest.mock('cozy-ui/transpiled/react/Spinner', () => {
  return ({ size }) => <div data-testid="Spinner" data-size={size}></div>
})
jest.mock('cozy-ui/transpiled/react/CircleButton', () => {
  return () => <div data-testid="CircleButton"></div>
})
jest.mock('cozy-ui/transpiled/react/I18n', () => {
  return { useI18n: () => ({ t: x => x }) }
})

describe('PermissionsApplication', () => {
  beforeEach(() => {
    const queryResult = {
      fetchStatus: 'loaded',
      data: {
        permissions: {
          files: {
            type: 'io.cozy.files',
            description: 'Required to access the files'
          },
          allFiles: {
            type: 'io.cozy.files.*',
            description: 'Required to access the files'
          },
          apps: {
            type: 'io.cozy.apps',
            description:
              'Required by the cozy-bar to display the icons of the apps',
            verbs: ['GET']
          },
          accounts: {
            type: 'io.cozy.accounts',
            description: 'Required to access accounts',
            verbs: ['POST']
          }
        }
      }
    }
    isQueryLoading.mockReturnValue(true)
    hasQueryBeenLoaded.mockReturnValue(true)
    Q.mockReturnValue({ getById: () => 'kfrf' })
    useQuery.mockReturnValue(queryResult)
    useClient.mockReturnValue({
      getStackClient: () => {
        return { uri: 'http://cozy.localhost:8080' }
      },
      getInstanceOptions: () => {
        return {
          subdomain: 'flat'
        }
      }
    })
  })
  it('should display slugName when query has been loaded', () => {
    hasQueryBeenLoaded.mockReturnValue(true)
    const { container } = render(<PermissionsApplication />)
    expect(container).toMatchSnapshot()
  })

  it('should render a spinner when query is loading and has not been loaded', () => {
    isQueryLoading.mockReturnValue(true)
    hasQueryBeenLoaded.mockReturnValue(false)
    const { queryByTestId } = render(<PermissionsApplication />)
    expect(queryByTestId('Spinner')).toBeTruthy()
  })

  it('should display slugName when query is not loading', () => {
    isQueryLoading.mockReturnValue(false)
    const { queryByText } = render(<PermissionsApplication />)
    expect(queryByText('DRIVE')).toBeInTheDocument()
  })

  it('should render Permissions.failedRequest when query status is failed', () => {
    useQuery.mockReturnValue({ fetchStatus: 'failed' })
    const { queryByText } = render(<PermissionsApplication />)
    expect(queryByText('Permissions.failedRequest')).toBeTruthy()
  })

  describe('completePermission', () => {
    it('should add description in permissions', () => {
      const key = 'contactsAccounts'
      const permission = 'Comptes associés aux contacts'
      const value = {
        description: 'blabla',
        verbs: ['GET', 'POST']
      }
      const resultat = completePermission(key, permission, value)
      expect(resultat).toEqual({
        name: 'contactsAccounts',
        title: 'Comptes associés aux contacts',
        description: 'blabla',
        verbs: ['GET', 'POST']
      })
    })
  })
})
