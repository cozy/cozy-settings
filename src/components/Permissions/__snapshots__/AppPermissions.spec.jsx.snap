// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionsApplication should display slugName when query has been loaded 1`] = `
<div>
  <div
    data-narrow="true"
    data-testid="page"
  >
    <div>
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root u-mr-half large"
        href="#/permissions"
        tabindex="0"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            class="styles__icon___23x3R"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M3.414 7h11.58C15.548 7 16 7.444 16 8c0 .552-.45 1-1.007 1H3.414l5.293 5.293a1 1 0 01-1.414 1.414l-7-7a1 1 0 010-1.414l7-7a1 1 0 111.414 1.414L3.414 7z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
      <div
        data-testid="NavigationList"
      >
        <div
          data-testid="pageTitle"
        >
          DRIVE
        </div>
        <div
          data-testid="CircleButton"
        />
        <div
          data-testid="NavigationListSection"
        >
          <div>
            <div
              data-testid="page"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root"
                >
                  <svg
                    class="styles__icon___23x3R"
                    height="24"
                    viewBox="0 0 48 48"
                    width="24"
                  >
                    <g
                      fill="none"
                      fill-rule="evenodd"
                    >
                      <path
                        d="M3 9c0 29.178 11.311 34.3 20.997 39C33.682 43.3 45 38.178 45 9L23.998 0 3 9z"
                        fill="#d6d8da"
                        fill-rule="nonzero"
                      />
                      <path
                        d="M22.5 25.244V30.5a1 1 0 001 1h1a1 1 0 001-1v-5.256A4.502 4.502 0 0024 16.5a4.5 4.5 0 00-1.5 8.744zM24 37.5c-8.284 0-15-6.716-15-15 0-8.284 6.716-15 15-15 8.284 0 15 6.716 15 15 0 8.284-6.716 15-15 15z"
                        fill="#5d6165"
                      />
                    </g>
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root MuiListItemText-multiline"
                >
                  <span
                    class="MuiTypography-root u-db u-ellipsis MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    CozyPermissions.Permissions.io.cozy.accounts
                  </span>
                  <span
                    class="MuiTypography-root u-db u-ellipsis MuiListItemText-secondary MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                  >
                    Permissions.write
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </div>
            <hr
              class="MuiDivider-root"
            />
          </div>
          <div>
            <div
              data-testid="page"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root"
                >
                  <svg
                    class="styles__icon___23x3R"
                    height="24"
                    viewBox="0 0 48 48"
                    width="24"
                  >
                    <g
                      fill="none"
                      fill-rule="evenodd"
                    >
                      <path
                        d="M25.342 24.67l20.87 10.436a1 1 0 010 1.788L25.341 47.33a3 3 0 01-2.684 0L1.788 36.894a1 1 0 010-1.788l20.87-10.435a3 3 0 012.684 0z"
                        fill="#d6d8da"
                      />
                      <path
                        d="M25.342 12.67l20.87 10.436a1 1 0 010 1.788L25.341 35.33a3 3 0 01-2.684 0L1.788 24.894a1 1 0 010-1.788l20.87-10.435a3 3 0 012.684 0z"
                        fill="#95999d"
                      />
                      <path
                        d="M25.342.67l20.87 10.436a1 1 0 010 1.788L25.341 23.33a3 3 0 01-2.684 0L1.788 12.894a1 1 0 010-1.788L22.659.67a3 3 0 012.684 0z"
                        fill="#5d6165"
                      />
                    </g>
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root MuiListItemText-multiline"
                >
                  <span
                    class="MuiTypography-root u-db u-ellipsis MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    CozyPermissions.Permissions.io.cozy.apps
                  </span>
                  <span
                    class="MuiTypography-root u-db u-ellipsis MuiListItemText-secondary MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                  >
                    Permissions.read
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </div>
            <hr
              class="MuiDivider-root"
            />
          </div>
          <div>
            <div
              data-testid="page"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root"
                >
                  <svg
                    class="styles__icon___23x3R"
                    height="24"
                    viewBox="0 0 48 48"
                    width="24"
                  >
                    <g
                      fill="none"
                      fill-rule="evenodd"
                    >
                      <rect
                        fill="#d6d8da"
                        height="39"
                        rx="2"
                        width="48"
                        y="6"
                      />
                      <path
                        d="M2 3h18.172a2 2 0 011.414.586l1.828 1.828A2 2 0 0024.828 6H45a3 3 0 013 3H24.828a2 2 0 00-1.414.586l-1.828 1.828a2 2 0 01-1.414.586H0V5a2 2 0 012-2z"
                        fill="#5d6165"
                      />
                    </g>
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root MuiListItemText-multiline"
                >
                  <span
                    class="MuiTypography-root u-db u-ellipsis MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    CozyPermissions.Permissions.io.cozy.files
                  </span>
                  <span
                    class="MuiTypography-root u-db u-ellipsis MuiListItemText-secondary MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                  >
                    Permissions.readAndWrite
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </div>
            <hr
              class="MuiDivider-root"
            />
          </div>
          <div>
            <div
              data-testid="page"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root"
                >
                  <svg
                    class="styles__icon___23x3R"
                    height="24"
                    viewBox="0 0 48 48"
                    width="24"
                  >
                    <g
                      fill="none"
                      fill-rule="evenodd"
                    >
                      <rect
                        fill="#d6d8da"
                        height="39"
                        rx="2"
                        width="48"
                        y="6"
                      />
                      <path
                        d="M2 3h18.172a2 2 0 011.414.586l1.828 1.828A2 2 0 0024.828 6H45a3 3 0 013 3H24.828a2 2 0 00-1.414.586l-1.828 1.828a2 2 0 01-1.414.586H0V5a2 2 0 012-2z"
                        fill="#5d6165"
                      />
                    </g>
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root MuiListItemText-multiline"
                >
                  <span
                    class="MuiTypography-root u-db u-ellipsis MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    CozyPermissions.Permissions.io.cozy.files.*
                  </span>
                  <span
                    class="MuiTypography-root u-db u-ellipsis MuiListItemText-secondary MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                  >
                    Permissions.readAndWrite
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </div>
            <hr
              class="MuiDivider-root"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
